apps/src/code-studio/reporting.js

```diff
@@ -39,13 +40,173 @@ reporting.getLastServerResponse = function () {
+        break;
+      case 'result':
+        if (inLevelGroup) {
+          // A multi in an assessment seems to send an object here instead of a
+          // boolean (which may well be a bug).
+          validateType('result', value, 'object');
+        } else {
+          validateType('result', value, 'boolean');
+        }
+        break;
+      case 'pass':
+        if (inLevelGroup) {
+          // A multi in an assessment seems to send an object here instead of a
+          // boolean (which may well be a bug).
+          validateType('pass', value, 'object');
+        } else {
+          validateType('pass', value, 'boolean');
+        }
+        break;
+      case 'testResult':
+        validateType('testResult', value, 'number');
+        break;
+      case 'submitted':
+        // In sendResultsCompletion this becomes either "true" the string  or false the boolean
+        // Would probably be better long term if it was always a string or always a boolean
+        if (value !== "true" && value !== false) {
+          console.error('Expected submitted to be either string "true" or value false');
+        }
+        break;
+      case 'onComplete':
+        if (value !== undefined) {
+          validateType('onComplete', value, 'function');
+        }
+        break;
+      case 'time':
+        validateType('time', value, 'number');
+        break;
+      case 'lines':
+        validateType('lines', value, 'number');
+        break;
+      case 'save_to_gallery':
+        validateType('save_to_gallery', value, 'boolean');
+        break;
+      case 'attempt':
+        validateType('attempt', value, 'number');
+        break;
+      case 'image':
+        if (value !== null) {
+          validateType('image', value, 'string');
+        }
+        break;
+      case 'containedLevelResultsInfo':
+        validateType('containedLevelResultsInfo', value, 'object');
+        break;
+      default:
+        // Eventually we'd probably prefer to throw here, but I don't have enough
+        // confidence that this validation is 100% correct to start breaking things
+        // if it isnt.
+        console.error(`Unexpected report key '${key}' of type '${typeof report[key]}'`);
+        break;
+    }
+  }
+}
+
 reporting.sendReport = function (report) {
+  // The list of report fields we want to send to the server
+  const serverFields = [
+    'program',
+    'app',
+    'allowMultipleSends',
+    'level',
+    'result',
+    'testResult',
+    'submitted',
+    'time',
+    'lines',
+    'save_to_gallery',
+    'attempt',
+    'image'
+  ];
+
+  validateReport(report);
+
   // jQuery can do this implicitly, but when url-encoding it, jQuery calls a method that
   // shows the result dialog immediately
   var queryItems = [];
   for (var key in report) {
-    if (report.hasOwnProperty(key) && key !== 'onComplete') {
-      queryItems.push(key + '=' + report[key]);
+    if (report.hasOwnProperty(key)) {
+      if (serverFields.includes(key)) {
```

Comment:
This seems like an appropriate place to use [`_.pick`](https://lodash.com/docs/4.17.4#pick)